### DISCUSSION TITLE
role resource permission document

### DISCUSSION THREAD

**Discussion by katsuharu** [argoproj/argo-workflows#10404] [Q&A]
Hello I’d like to a question.

I want to set Argo Workflow permission via Service Account.

I search a official document about argo workflow role document, I couldn’t.
Is there any document?( resources, verbs)
I want to to know what verb does and not on each resource.

ex)  what operation verb "patch" can for resource "workflows"
https://github.com/argoproj/argo-workflows/blob/master/manifests/namespace-install/argo-server-rbac/argo-server-role.yaml

ex) like k8s api document
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/

relevant below discussion
https://github.com/argoproj/argo-workflows/discussions/7456

**Comment by wesleyscholl** [comment]
@katsuharu 

The Argo Workflows `patch` operation allows updates to `Workflow` objects. This operation can be used to modify `Workflow` specifications dynamically. Some examples include adding or updating labels and annotations, or changing workflow spec configurations. Other examples:

- **Pod State Updates**: Argo Workflows uses the `patch` action to modify the workflow's status field (i.e. recording task progress and errors). This is critical for workflows that require outputs or status tracking.

- **PodSpec Modifications**: The `podSpecPatch` field in a workflow spec can be used to dynamically adjust pod resource requests or limits. This is useful for handling large artifacts or specific memory and CPU requirements without modifying the rest of the workflow template.

- **Customizing Artifacts**: `patch` can be used to alter artifact behaviors, such as modifying compression settings or workflow artifact garbage collection strategies.

- **Garbage Collection Strategies**: `patch` enables modification or fine-tuning artifact garbage collection settings to include artifact retention or applying default cleanup strategies.

### Argo Workflows Resources and Permissions Table

| **Resource**                 | **API Group**                  | **Verbs**                                | **Description**                                                                 |
|-------------------------------|--------------------------------|------------------------------------------|---------------------------------------------------------------------------------|
| `Workflow`                   | `argoproj.io`                 | `get`, `list`, `watch`, `create`, `delete`, `update`, `patch` | Manage workflows, including creating, updating, and deleting workflows.         |
| `WorkflowTemplate`           | `argoproj.io`                 | `get`, `list`, `watch`, `create`, `delete` | Manage reusable workflow templates.                                             |
| `ClusterWorkflowTemplate`    | `argoproj.io`                 | `get`, `list`, `watch`, `create`, `delete` | Manage workflow templates at the cluster level.                                 |
| `CronWorkflow`               | `argoproj.io`                 | `get`, `list`, `watch`, `create`, `delete`, `update` | Manage scheduled workflows.                                                     |
| `Pod`                        | `core`                       | `get`, `list`, `watch`, `create`, `delete` | Manage workflow-related pods.                                                   |
| `Service`                    | `core`                       | `get`, `list`, `watch`, `create`, `delete` | Manage Kubernetes services related to workflows.                                |
| `ConfigMap`                  | `core`                       | `get`, `list`, `watch`, `create`, `delete`, `update` | Manage workflow configurations.                                                 |
| `Secret`                     | `core`                       | `get`, `create`, `update`, `delete`      | Access sensitive information used in workflows.                                 |
| `Event`                      | `core`                       | `get`, `list`, `watch`, `create`, `delete` | Track workflow events.                                                          |
| `PersistentVolumeClaim`      | `core`                       | `get`, `list`, `watch`, `create`, `delete` | Manage storage volumes used in workflows.                                       |
| `Role`                       | `rbac.authorization.k8s.io`  | `get`, `create`, `update`, `delete`, `bind` | Manage role-based access control.                                               |
| `RoleBinding`                | `rbac.authorization.k8s.io`  | `get`, `create`, `update`, `delete`, `bind` | Bind roles to users or groups.                                                  |
| `ClusterRole`                | `rbac.authorization.k8s.io`  | `get`, `create`, `update`, `delete`, `bind` | Manage cluster-wide role-based access control.                                  |
| `ClusterRoleBinding`         | `rbac.authorization.k8s.io`  | `get`, `create`, `update`, `delete`, `bind` | Bind cluster roles to users or groups.                                          |
| `Job`                        | `batch`                      | `get`, `list`, `watch`, `create`, `delete`, `update` | Manage batch jobs executed by workflows.                                        |
| `Deployment`                 | `apps`                       | `get`, `list`, `watch`, `create`, `delete`, `update` | Manage workflow-related deployments.                                            |


### Argo Workflows Verbs and Descriptions Table

| **Verb**   | **Description**                                                                 |
|------------|---------------------------------------------------------------------------------|
| `get`      | Retrieve details of a resource.                                                |
| `list`     | List multiple resources of the same type.                                      |
| `watch`    | Monitor changes to resources in real-time.                                     |
| `create`   | Create a new resource.                                                         |
| `delete`   | Delete an existing resource.                                                   |
| `update`   | Modify an existing resource.                                                   |
| `patch`    | Apply partial updates to a resource.                                           |
| `bind`     | Assign roles to users or groups (specific to RBAC resources).                  |


### Resources:
1. [Kubernetes RBAC API Documentation](https://kubernetes.io/docs/reference/access-authn-authz/rbac/)
2. [Argo Workflows Service Accounts](https://argo-workflows.readthedocs.io/en/latest/service-accounts/)
3. [Argo Workflows SSO RBAC](https://argo-workflows.readthedocs.io/en/latest/argo-server-sso/#sso-rbac)
4. [Argo Server RBAC Role Example](https://github.com/argoproj/argo-workflows/blob/master/manifests/namespace-install/argo-server-rbac/argo-server-role.yaml)
5. [Argo Workflows Workflow RBAC](https://argo-workflows.readthedocs.io/en/latest/workflow-rbac/)
6. [Argo Workflows Security](https://argo-workflows.readthedocs.io/en/latest/security/)
7. [Argo Workflows Workflow Manager Role](https://github.com/argoproj/argo-workflows/blob/main/manifests/quick-start/base/workflow-manager-role.yaml)
8. [Argo Workflows Workflow Controller Role](https://github.com/argoproj/argo-workflows/blob/main/manifests/cluster-install/workflow-controller-rbac/workflow-controller-role.yaml)
9. [Argo Workflows Agent Role](https://github.com/argoproj/argo-workflows/blob/main/manifests/quick-start/base/agent-role.yaml)
